### PR TITLE
Disable Xdebug by default unless specified by user

### PR DESCRIPTION
**Pull request by brylie**
## Description

Currently, `wp-env` installs Xdebug regardless of whether the user has requested it. The only installation check is based on PHP version being 7.2+.

Xdebug shouldn't be installed unless specified by the user. Likewise, the Docker file should be more careful when running the `pecl install xdebug`  command, as [noted by @noahtallen](https://github.com/WordPress/gutenberg/pull/35667#issuecomment-953278245). 

The added complexity of the Xdebug installation is causing some other issues:

- https://github.com/WordPress/gutenberg/issues/30213
- https://github.com/WordPress/gutenberg/issues/27783
- https://github.com/WordPress/gutenberg/issues/34320

## Checklist:
- [x] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by derickr** [comment]
It's spelled "Xdebug" (no capital "D").

> XDebug shouldn't be installed unless specified by the user.

FWIW, I don't think this is wise, as you're going to make it harder for people to get a working debugging setup.

**Comment by brylie** [comment]
> It's spelled "Xdebug" (no capital "D").

Thanks for pointing that out. I corrected the title and description :smiley: 

> FWIW, I don't think this is wise, as you're going to make it harder for people to get a working debugging setup.

From what I understand `config.xdebug` defaults to `off`.

This pull request avoids automatically installing Xdebug when `config.xdebug === 'off'`. When the developer specifies an `xdebug` value other than `off`, Xdebug will be installed.

It may be worth pointing out that `wp-env`/Docker will re-build the development container when enabling Xdebug, so it may be OK to omit the Xdebug installation step in cases where it isn't needed.

**Comment by brylie** [comment]
There is currently no way to disable the Xdebug installation, even when it is desirable to do so. I am encountering a bug during the Xdebug installation (#34320) step and would just like to bypass the Xdebug installation to continue with my development task.

image

For simplicity's sake, it is important only to install dependencies that are needed.

**Comment by koengabriels** [comment]
hope this gets through because I am working for a client that uses wp-env to set up unit tests on their precommit hook
since I am getting this error and have been unable to bypass it... I can not commit anything :(

**Comment by brylie** [comment]
@koengabriels would you mind taking a close look at the code and possibly testing it on your end? Your peer review would be much appreciated :smiley:

**Comment by koengabriels** [comment]
@brylie I have to look into how to do that, I have zero experience with testing a PR from Github or wp-env code base 
if I can I will post feedback here, appreciate the time you took to create a pull request for this issue :slightly_smiling_face:

File: packages/env/lib/init-config.js [hunk 1]
```diff
@@ -107,21 +107,34 @@ module.exports = async function initConfig( {
...
+			}
+
+			// Xdebug 3 supports 7.2 and higher.
+			// Enable Xdebug for PHP >= 7.2.
+			const usingCompatiblePhpVersion = majorVersion > 7 && minorVersion >= 2;
+
+			if ( usingCompatiblePhpVersion ) {
+				shouldInstallXdebug = true;
+			} else {
+				throw new Error( 'Cannot use XDebug 3 on PHP < 7.2.' );
```
Comment (noahtallen) [inline packages/env/lib/init-config.js]: Nice improvement. As a result, if someone runs `wp-env start --xdebug` with an incorrect php version, they'll get a helpful error message, but plain wp-env start will continue working.

File: packages/env/lib/init-config.js [hunk 2]
```diff
@@ -107,21 +107,34 @@ module.exports = async function initConfig( {
...
+			}
+		} else {
+			// By default, an undefined phpVersion uses the version on the docker image,
+			// which is supported by Xdebug 3.
+			shouldInstallXdebug = true;
```
Comment (noahtallen) [inline packages/env/lib/init-config.js:152]: we could phrase it like this to remove some of the duplicated conditionals:  

```js
if ( config.xdebug.... ) {
  if ( ...php...) {
    ...
    if( majorVersion < 7 || ( majorVersion === 7 && minorVersion < 2 ) ) {
      throw new Error()
    }
  }
  shouldInstallXdebug = true;
} 
```

File: packages/env/lib/init-config.js [hunk 3]
```diff
@@ -107,21 +107,34 @@ module.exports = async function initConfig( {
...
+			const majorVersion = parseInt( versionTokens[ 0 ] );
+			const minorVersion = parseInt( versionTokens[ 1 ] );
+
+			if ( isNaN( majorVersion ) || isNaN( minorVersion ) ) {
+				throw new Error( 'Something went wrong parsing the php version.' );
+			}
+
+			// Xdebug 3 supports 7.2 and higher.
+			// Enable Xdebug for PHP >= 7.2.
+			const usingCompatiblePhpVersion = majorVersion > 7 && minorVersion >= 2;
```
Comment (noahtallen) [inline packages/env/lib/init-config.js]: I think we should keep the original conditional; this fails in a couple conditions:

- doesn't support php 7
- doesn't support php x.1 or x.0 if the major version is supported

**Review by noahtallen — COMMENTED** [review]
Left a couple comments, but I think we can move forward with this. 

This doesn't change the public API for using xdebug with wp-env, so it isn't any harder to enable with wp-env. :+1: 

I tested locally, and it will create a wordpress image without xdebug by default, and one with xdebug as soon as the `--xdebug` flag is passed.

Another note is to run prettier locally to get the lint check passing :+1:

**Comment by brylie** [comment]
@noahtallen, I've incorporated your suggested changes to the nested logic and conditional as well as the suggestion from PR #35667

Are there any other changes you would recommend?

File: packages/env/lib/init-config.js (hunk 2 above)
Comment (brylie) [reply packages/env/lib/init-config.js:152]: I've just committed this change to the nested logic. Please let me know if I've misunderstood the suggestion.

File: packages/env/lib/init-config.js (hunk 3 above)
Comment (brylie) [reply packages/env/lib/init-config.js]: I just committed the changes you suggested in a previous comment, including the desired conditional:

```
if( majorVersion < 7 || ( majorVersion === 7 && minorVersion < 2 ) )
```

Comment (brylie) [reply packages/env/lib/init-config.js]: Please let me know if I've misunderstood the suggestion :smile:

**Review by noahtallen — APPROVED** [review]
This tests well for me, thank you for implementing the suggestions!

I don't have access to merge unless the CI checks are passing. I'd recommend running prettier/eslint again and rebasing the PR on the latest trunk changes.

It would also be great if you could add a line to the changelog (https://github.com/WordPress/gutenberg/blob/trunk/packages/env/CHANGELOG.md) under the "unreleased" section (probably under a new heading `### Enhancement`)

Other than that, I think this is good to go :+1:

**Comment by brylie** [comment]
Sorry for the review request bonanza. I tried to rebase changes from `trunk` into this branch so the CI tasks would succeed.

Would someone with admin rights please help bypass the CI review since most tasks are unrelated to this pull request?

@noahtallen, would you please review the recent changes to this PR? I've refactored the PHP compatibility check into its own function so the `dockerFileContents` function is single-purpose.

**Comment by swissspidy** [comment]
Why bypass CI review?

There are some actual lint errors in `packages/env/lib/init-config.js` that need to get addressed.

**Comment by brylie** [comment]
> Another note is to run prettier locally to get the lint check passing

When I run `npm run format-js` locally, I get the following error:

```
$ npm run format-js

> gutenberg@11.8.0 format-js
> wp-scripts format-js

sh: 1: wp-scripts: not found
```

When I have tried running `npm install`, I get many errors related to `Tracker "idealTree:inflate:s" already exists`

[2021-10-29T08_01_35_754Z-debug.log.txt](https://github.com/WordPress/gutenberg/files/7440050/2021-10-29T08_01_35_754Z-debug.log.txt)

I believe these issues are related to the monorepo nature of this project whereby it is necessary to install the entire Gutenberg development stack as well as any other libraries published by `@wordpress` in order to work on `wp-env`

#34325

**Comment by swissspidy** [comment]
Yes it's a monorepo and you'd need to run `npm install` from the repo root.

You should also be using the latest Node LTS, Node 16 (run `nvm install` to fix). According to your logs you are still on Node 12

**Comment by brylie** [comment]
I ran `nvm install 16 && nvm use 16`:
image


Running `npm install` in the monorepo root seems to have produced pretty much the same errors:

[2021-10-29T08_48_38_880Z-debug.log.txt](https://github.com/WordPress/gutenberg/files/7440406/2021-10-29T08_48_38_880Z-debug.log.txt)

**Comment by brylie** [comment]
@noahtallen, it seems like the CI checks are all passing or skipped. Would you mind giving this another review?

**Comment by noahtallen** [comment]
> I ran nvm install 16 && nvm use 16:

For future reference, I think this is because gutenberg still requires [npm v6](https://github.com/WordPress/gutenberg/blob/211eebe72d84a13da1f7d0fbefc08329aed1040b/package.json#L19). I think there is a project to update that, though.

But now that CI is passing, it shouldn't be a problem for this PR :)

**Review by noahtallen — APPROVED** [review]
The latest changes also test well, thank you!

This change will get released to npm whenever the gutenberg project does its next npm release, though I'm not sure when that will be unfortunately. I think the schedule changes around WordPress release windows

**Pull request merged**